backend/tracim_backend/lib/core/url_preview.py

```diff
@@ -39,25 +44,69 @@ def __init__(self, title: Optional[str], description: Optional[str], image: Opti
         self.image = image
 
 
+class URLPreviewResponse(LengthLimitedResponse):
+    def __init__(self, response: LengthLimitedResponse) -> None:
+        self.__setstate__(response.__getstate__())
+        self.content_length_limit = response.content_length_limit
+        self.origin_url = response.origin_url
+
+    @property
+    def is_html(self) -> bool:
+        return self.headers.get("content-type").startswith("text/html")
+
+    @property
+    def is_image(self) -> bool:
+        return self.headers.get("content-type").startswith("image")
+
+    @property
+    def filename(self) -> str:
+        # TODO - G.M - 2022-02-16 - use content-disposition header if exist instead
+        # of url path.
+        # this is tricky as we need to decode it if utf-8 encoded.
+        return urlparse(self.url).path.rsplit("/", 1)[-1]
+
+    @property
+    def mimetype(self) -> str:
+        return self.headers.get("content-type")
+
+
 class URLPreviewLib(object):
     def __init__(self, config: CFG,) -> None:
         self.app_config = config
 
     def get_preview(self, url: str) -> URLPreview:
-
         try:
-            title, description, image_url = web_preview(
-                url,
-                timeout=self.app_config.URL_PREVIEW__FETCH_TIMEOUT,
-                headers={
-                    "User-Agent": "Mozilla/5.0 (compatible; Tracim bot; +https://www.tracim.fr)",
-                    "Accept": "image/*,text/html;q=0.8",
-                    # INFO - SG - 2021-04-16 - Needed to bypass google and youtube cookie consent banner
-                    # See https://github.com/tracim/tracim/issues/4470
-                    "Cookie": "CONSENT=PENDING+999",
-                },
+            response = URLPreviewResponse(
+                do_request(
+                    url,
+                    timeout=self.app_config.URL_PREVIEW__FETCH_TIMEOUT,
+                    headers={
+                        "User-Agent": "Mozilla/5.0 (compatible; Tracim bot; +https://www.tracim.fr)",
+                        "Accept": "image/*,text/html;q=0.8",
+                        # INFO - SG - 2021-04-16 - Needed to bypass google and youtube cookie consent banner
+                        # See https://github.com/tracim/tracim/issues/4470
+                        "Cookie": "CONSENT=PENDING+999",
+                    },
+                    content_length_limit=MAX_CONTENT_LENGTH_FOR_URL_PREVIEW,
+                )
             )
+            # INFO - GM - 2022-02-22 - Default case: preview from html metadata
+            if response.is_html:
+                title, description, image_url = web_preview(url=url, content=response.text_content)
```

Comment:
Here we still make two requests, wouldn't adding the possibility to give the content to the `web_preview` function be possible? So that this function becomes:
1. build the request
2. make the checks (content-length/…)
3. depending on the check call web_preview or directly return/adapt the image if it is one